### ISSUE TITLE
effect.code is run before effect's fields are initialized (test stdin fails in c-backend when using field instead of literal)

### ISSUE THREAD

**Issue by michaellilltokiwa** [tokiwa-software/fuzion#317]
In io.in.stdin.readline:

Changing the line from this
`.takeWhile(b -> b != 10)`
to
`.takeWhile(b -> b != line_feed)` 
leads to test failure in c-backend.

(`line_feed` is a field `private line_feed u8 := 10`)

**Comment by fridis** [comment]
The problem is that an effect like `io.stdin` gets installed by call inheritance-call `effect r`, which, for ref instances, installs a copy of that instance in the current thread. This code, however, is executed before the code for `io.stdin` is executed, and hence the field `io.stdin.line_feed` is not initialized yet. 

It works in the interpreter due to a bug in the interpreter: value type effects, when installed, do not get copied, but a ref to the original value is stored. Consequently, the later initialization of `io.stdin.line_feed` also affects the installed effect which is a reference to the same instance. 

To solve this, we will have to 
* fix the interpreter to create copies of value type effects when installed
* implement static analysis to report the use of an uninitialized field
* fix the problem, e.g., by turing `io.stdin`into a `ref` or by finding a better way to install an effect only after it was intialized completely.

**Comment by fridis** [comment]
Here is a small example that illustrates the problem when interpreted and when compiled: 

```
 > cat ex317.fz
buggyEffect is

  e(code ()->unit) : effect is
    x := 123
    say "e.x initialized"
    action => say x

  x := e ()->run
  run => e.env.action
  x.action
  unit
 > ./build/bin/fz ex317.fz
-555555555
e.x initialized
123
 > ./build/bin/fz -c ex317.fz
 > ./buggyEffect 
0
e.x initialized
123
```

**Comment by fridis** [comment]
The problem is that an `effect`'s constructor installs the `effect` and runs the `code` that was provided before the constructor of the feature that inherits from `effect` is executed. If fields in the heir are accessed in any operations of the effect, these fields will not be initalized. 
A solution would be
- insted of running `code` by `effect`'s constructor, add a feature `use code` that installs the effect, runs `code` and uninstalls the effect.
- it would be nice to have some form of linear type to ensure that an instance of a sub-feature of `effect` can only be used for a call to `use`.
